src/metabase/api/advanced_computation/public.clj

```diff
@@ -6,14 +6,110 @@
             [metabase.api.advanced-computation.common :as advcomp-common]
             [metabase.api.card :as api.card]
             [metabase.api.common :as api]
+            [metabase.api.embed :as api.embed]
+            [metabase.api.preview-embed :as api.preview]
             [metabase.api.public :as api.public]
             [metabase.async.util :as async.u]
             [metabase.models.card :as card :refer [Card]]
+            [metabase.models.dashboard :as dashboard :refer [Dashboard]]
             [metabase.query-processor.streaming :as qp.streaming]
+            [metabase.util :as u]
+            [metabase.util.embed :as eu]
             [metabase.util.schema :as su]
             [schema.core :as s]
             [toucan.db :as db]))
 
+;;
+;; helper functions
+;;
+
+(defn run-query-for-card-with-id-async
+  "Run the query belonging to Card with `card-id` with `parameters` and other query options (e.g. `:constraints`).
+  Returns a `StreamingResponse` object that should be returned as the result of an API endpoint."
+  [card-id export-format parameters & options]
+  {:pre [(integer? card-id)]}
+  ;; run this query with full superuser perms
+  ;;
+  ;; we actually need to bind the current user perms here twice, once so `card-api` will have the full perms when it
+  ;; tries to do the `read-check`, and a second time for when the query is ran (async) so the QP middleware will have
+  ;; the correct perms
+  (binding [api/*current-user-permissions-set* (atom #{"/"})]
+    (apply api.card/run-query-for-card-async card-id export-format
+           :parameters parameters
+           :context    :public-question
+           :run        (fn [query info]
+                         (qp.streaming/streaming-response [{:keys [reducedf], :as context} export-format]
+                           (advcomp-common/check-query-type :query query)
+                           (let [context  (assoc context :reducedf (api.public/public-reducedf reducedf))
+                                 in-chan  (binding [api/*current-user-permissions-set* (atom #{"/"})]
+                                            (advcomp-common/run-query context query info))
+                                 out-chan (a/promise-chan (map api.public/transform-results))]
+                             (async.u/promise-pipe in-chan out-chan)
+                             out-chan)))
+           options)))
```

Comment:
this is 95% the same as the version in `metabase.api.public` -- only the implementation of `:run` is any different. The reason `:run` is a param to the original version is so you can override it if needed. So you can toss most of the stuff here and just use the original version and override just `:run`.

Out of the other 10 lines in `:run`, the only differences are the extra check and using a different function to run the query. It's pretty easy to find a way to reuse that stuff too. If you change the original version to

```clj
(defn default-run-query-for-card-with-id-async-run-fn
  [export-format & {:keys [run-fn]
                    :or   {run-fn qp/process-query-and-save-execution!}}]
  (fn [query info]
    (qp.streaming/streaming-response [{:keys [reducedf], :as context} export-format]
      (let [context  (assoc context :reducedf (public-reducedf reducedf))
            in-chan  (binding [api/*current-user-permissions-set* (atom #{"/"})]
                       (run-fn query info context))
            out-chan (a/promise-chan (map transform-results))]
        (async.u/promise-pipe in-chan out-chan)
        out-chan))))

(s/defn run-query-for-card-with-id-async
  "Run the query belonging to Card with `card-id` with `parameters` and other query options (e.g. `:constraints`).
  Returns a `StreamingResponse` object that should be returned as the result of an API endpoint."
  [card-id export-format parameters & options]
  {:pre [(integer? card-id)]}
  ;; run this query with full superuser perms
  ;;
  ;; we actually need to bind the current user perms here twice, once so `card-api` will have the full perms when it
  ;; tries to do the `read-check`, and a second time for when the query is ran (async) so the QP middleware will have
  ;; the correct perms
  (binding [api/*current-user-permissions-set* (atom #{"/"})]
    (apply card-api/run-query-for-card-async card-id export-format
           :parameters parameters
           :context    :public-question
           :run        (default-run-query-for-card-with-id-async-run-fn export-format)
           options)))
```

Then your new version can be

```clj
(defn run-pivot-table-query-for-card-with-id-async
  [card-id export-format parameters & options]
  (run-query-for-card-with-id-async
   :run (let [run-fn (default-run-query-for-card-with-id-async-run-fn
                      export-format
                      :run-fn advcomp-common/run-query)]
          (fn [query info]
            (advcomp-common/check-query-type :query query)
            (run-fn query info)))))
```

That's a lot nicer than copying a 25-line function to change 2 lines, and it means other people can benefit from the extra flexibility going forward